openmp on Travis fails for CLang and/or OSX

I've enabled OPENMP testing on Travis. We get the following type of run-time error
```
lm_to_projdata: error while loading shared libraries: libomp.so: cannot open shared object file: No such file or directory
```
on the CLang Linux and OSX builds, see https://travis-ci.org/UCL/STIR/builds/289769526 for all builds, and the [log of one of the failed ones](https://s3.amazonaws.com/archive.travis-ci.org/jobs/289769528/log.txt?X-Amz-Expires=30&X-Amz-Date=20171019T072634Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20171019/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=fc4801cb7d8a7100384d36223adbde72916513abc07eb71cc8b58e8c29a45b57). Linking is fine. gcc builds are fine as well.

Latest results with Travis (after merge with master):
- Linux with gcc is ok
- Linux with clang: List of linking errors with e.g. `__kmpc_global_thread_num`
- OSX with clang and gcc:
```
 CMake Error at /usr/local/Cellar/cmake/3.10.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
```

I don't have much time at this moment,
but if the problem still exists in April, then I promise to fix it.

But in the meantime:
There are two challenges with openmp using clang:
-it requires different command line options both for the linker and the compiler (and it needs a different library to be linked)
- old clangs did not have proper openmp support, and travis might just use a too old.

Possible fixes:
- cmake: 
  - disable/warn openmp for (old) clang
  - add clang specific options
- travis: 
  - skip openmp for clang
  - update clang/install proper openmp lib

See: https://stackoverflow.com/questions/33357029/using-openmp-with-clang

Best,
  David

On February 2, 2018 7:45:31 PM GMT+01:00, Kris Thielemans <notifications@github.com> wrote:
>Latest results with Travis (after merge with master):
>- Linux with gcc is ok
>- Linux with clang: List of linking errors with e.g.
>`__kmpc_global_thread_num`
>- OSX with clang and gcc:
>```
>CMake Error at
>/usr/local/Cellar/cmake/3.10.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137
>(message):
>  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
>```
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/UCL/STIR/issues/117#issuecomment-362669021

I am trying to install OpenMP on my Mac, and I’m having problems.

I have been using a fully declared path as /usr/local/opt/llvm/bin/clang (or clang++ for cxx_compiler) It configures ok but when it generates it says:

CMake Error in src/recon_buildblock/CMakeLists.txt:
No known features for CXX compiler
"Clang"
version 5.0.2.

I tried with different versions of clang and I have the same problem.

Hi, going to give a rundown of what I did that resulted in an OpenMP STIR on OSX.

> presumably fixed by #413

I believe this is now fixed, I can configure with the llvm clang on mac (installed with `brew install llvm`). I setup a new build directory and use the command.
```
ccmake ${STIR_SOURCE} -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang
```

When setting STIR_OpenMP to True, configuing gave the following error:
```
 CMake Error at /usr/local/Cellar/cmake/3.17.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
   Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
```

----
Setting up the cmake, the setting mentioned by Richard:
>Then, I set STIR_OpenMP to true, and had to set the following four options:
>- OpenMP_CXX_FLAGS=-fopenmp
>- OpenMP_C_FLAGS=-fopenmp
>- OpenMP_CXX_LIB_NAMES= (delete the NOTFOUND)
>- OpenMP_C_LIB_NAMES= (delete the NOTFOUND)

did not work, but the exports did:
>`export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"`
> `export LDFLAGS="-L/usr/local/opt/llvm/lib`

and the OMP specific libraries and flags were set automatically by cmake. STIR installed with ROOT, ITK, HDF5, and python 3.8 support and the recon_test_pack tests ran without fault, including `run_root_GATE.sh`. Testing OSMAPOSL resulted in the use of all 8 cores on my machine.

--- a/.travis.yml
+++ b/.travis.yml
@@ -104,11 +104,14 @@ before_install:
       #now upgrade to avoid problems with outdated packages
       brew upgrade
 
-      brew install llvm
-      brew install libomp
+      # check if we're using LLVM
+      if [[ $EXTRA_BUILD_FLAGS == *"llvm"* ]]; then
+        brew install llvm
+        brew install libomp
+        export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"
+        export LDFLAGS="-L/usr/local/opt/llvm/lib"
+      fi
 
-      export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"
-      export LDFLAGS="-L/usr/local/opt/llvm/lib"
 
       brew install ccache
       export PATH="/usr/local/opt/ccache/libexec:$PATH"

--- a/.travis.yml
+++ b/.travis.yml
@@ -45,7 +45,7 @@ matrix:
  - os: osx
    osx_image: xcode11.3
    python: 3
-   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=OFF -DSTIR_ENABLE_EXPERIMENTAL:BOOL=ON" CC=clang CXX=clang++
+   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=ON -DSTIR_ENABLE_EXPERIMENTAL:BOOL=ON" CC=clang CXX=clang++
 
 env:
   global:
@@ -97,6 +97,9 @@ before_install:
       brew update
       #now upgrade to avoid problems with outdated packages
       brew upgrade
+
+      export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"
+      export LDFLAGS="-L/usr/local/opt/llvm/lib
 
       brew install ccache
       export PATH="/usr/local/opt/ccache/libexec:$PATH"

--- a/.travis.yml
+++ b/.travis.yml
@@ -45,7 +45,13 @@ matrix:
  - os: osx
    osx_image: xcode11.3
    python: 3
-   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=ON -DSTIR_ENABLE_EXPERIMENTAL:BOOL=ON" CC=clang CXX=clang++
+   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=OFF -DSTIR_ENABLE_EXPERIMENTAL:BOOL=ON" CC=clang CXX=clang++
+
+   # osx, OpenMP using llvm clang
+ - os: osx
+   osx_image: xcode11.3
+   python: 3
+   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=ON -DSTIR_ENABLE_EXPERIMENTAL:BOOL=OFF -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang"
 
 env:
   global:
@@ -97,6 +103,9 @@ before_install:
       brew update
       #now upgrade to avoid problems with outdated packages
       brew upgrade
+
+      brew install llvm
+      brew install libomp
 
       export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"
       export LDFLAGS="-L/usr/local/opt/llvm/lib"

--- a/.travis.yml
+++ b/.travis.yml
@@ -99,7 +99,7 @@ before_install:
       brew upgrade
 
       export DYLD_FALLBACK_LIBRARY_PATH=$DYLD_FALLBACK_LIBRARY_PATH:"/usr/local/opt/llvm/lib"
-      export LDFLAGS="-L/usr/local/opt/llvm/lib
+      export LDFLAGS="-L/usr/local/opt/llvm/lib"
 
       brew install ccache
       export PATH="/usr/local/opt/ccache/libexec:$PATH"

--- a/.travis.yml
+++ b/.travis.yml
@@ -51,7 +51,7 @@ matrix:
  - os: osx
    osx_image: xcode11.3
    python: 3
-   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=ON -DSTIR_ENABLE_EXPERIMENTAL:BOOL=OFF -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang"
+   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=1 -DSTIR_OPENMP:BOOL=ON -DSTIR_ENABLE_EXPERIMENTAL:BOOL=ON -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang"
 
 env:
   global:

